benchmark/algorithms/httpann.py

```diff
@@ -0,0 +1,265 @@
+    | POST   | /set_query_arguments | dictionary of query arguments                                                                              | 200             | { }                                                                        |
+    | POST   | /query               | { "X": <query vectors as list of lists of floats>, "k": <number of neighbors to find and keep in memory> } | 200             | { }                                                                        |
+    | POST   | /range_query         | { "X": <query vectors as list of lists of floats>, “radius”: <distance to neighbors> }                     | 200             | { }                                                                        |
+    | POST   | /get_results         | { }                                                                                                        | 200             | { "get_results": <neighbors as list of lists of ints> }                    |
+    | POST   | /get_additional      | { }                                                                                                        | 200             | { "get_additional": <dictionary of arbitrary additional result metadata> } |
+    | POST   | /get_range_results   | { }                                                                                                        | 200             | { "get_range_results": <list of three 1-dimensional lists (lims, I, D)> }  |
+
+    Note that this is a 1:1 copy of the BaseANN Python Class API implemented as remote procedure calls.
+    """
+
+    def __init__(self, server_url: str, start_seconds: int, name: str, **kwargs):
+        self.server_url = server_url
+        self.name = name
+
+        # Used by get_results method, defined in query method.
+        self.res = []
+
+        # Let the server start and post to init.
+        time.sleep(start_seconds)
+        self.post("init", kwargs, 200)
+
+    def post(self, path: str, body: dict, expected_status: int) -> dict:
+        url = f"{self.server_url}/{path}"
+        res = requests.post(url, json=body)
+        if res.status_code != expected_status:
+            raise HttpANNResponseError(url, expected_status, res.status_code)
+        return res.json()
+
+    def fit(self, dataset):
+        body = dict(dataset=dataset)
+        self.post("fit", body, 200)
+
+    def load_index(self, dataset):
+        body = dict(dataset=dataset)
+        json = self.post("load_index", body, 200)
+        return json["load_index"]
+
+    def query(self, X, k):
+        body = dict(X=[arr.tolist() for arr in X], k=k)
+        self.post("query", body, 200)
+
+    def range_query(self, X, radius):
+        body = dict(X=[arr.tolist() for arr in X], radius=radius)
+        self.post("range_query", body, 200)
```

Comment:
I'm wondering what the performance penalty of this will be. I would be happy to 

- change the arguments such that the query vector file is exposed
- or add some kind of `prepare` for providing the query vectors

Reply:
I've done some local testing, and I think it will be fine since the HTTP overhead and JSON serialization is only incurred one time. It was a bigger deal with ann-benchmarks because that framework required a request and serialization for every query.